cmd/kubeadm/app/master/selfhosted.go

```diff
@@ -69,9 +87,271 @@ func CreateSelfHostedControlPlane(cfg *kubeadmapi.MasterConfiguration, client *c
+		cluster := &spec.Cluster{}
+
+		err := restClient.Get().
+			Resource(spec.TPRKindPlural).
+			Namespace(metav1.NamespaceSystem).
+			Name(etcdCluster).
+			Do().Into(cluster)
+
+		if err != nil {
+			fmt.Printf("[self-hosted] Error retrieving etcd cluster: %v\n", err)
+			return false, nil
+		}
+
+		switch cluster.Status.Phase {
+		case spec.ClusterPhaseRunning:
+			return true, nil
+		case spec.ClusterPhaseFailed:
+			return false, errors.New("failed to create etcd cluster")
+		default:
+			return false, nil
+		}
+	})
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Waiting for etcd to remove peer")
+	err = waitBootEtcdRemoved(cfg.Etcd.Cluster.ServiceIP)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Removing seed etcd")
+	// remove seed etcd pod
+	etcdStaticManifestPath := buildStaticManifestFilepath(etcd)
+	if err := os.RemoveAll(etcdStaticManifestPath); err != nil {
+		return fmt.Errorf("unable to delete seed etcd manifest [%v]", err)
+	}
+
+	fmt.Println("Waiting for seed etcd pod to delete")
+
+	//wait for seed etcd to disappear
+	wait.PollInfinite(kubeadmconstants.DiscoveryRetryInterval, func() (bool, error) {
+		_, err := client.Core().Pods(metav1.NamespaceSystem).Get(etcd, metav1.GetOptions{})
+		if err != nil && apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	})
+
+	fmt.Printf("[self-hosted] self-hosted etcd ready after %f seconds\n", time.Since(start).Seconds())
 	return nil
 }
 
+func waitBootEtcdRemoved(etcdServiceIP string) error {
+	err := wait.Poll(10*time.Second, 5*time.Minute, func() (bool, error) {
+		cfg := clientv3.Config{
+			Endpoints:   []string{fmt.Sprintf("http://%s:2379", etcdServiceIP)},
+			DialTimeout: 5 * time.Second,
+		}
+		etcdcli, err := clientv3.New(cfg)
+		if err != nil {
+			fmt.Printf("failed to create etcd client, will retry: %v", err)
+			return false, nil
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		m, err := etcdcli.MemberList(ctx)
+		cancel()
+		etcdcli.Close()
+		if err != nil {
+			fmt.Printf("failed to list etcd members, will retry: %v", err)
+			return false, nil
+		}
+
+		if len(m.Members) != 1 {
+			fmt.Println("still waiting for boot-etcd to be deleted...")
+			return false, nil
+		}
+		return true, nil
+	})
+	return err
+}
+
+func createEtcdService(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	etcdService := getEtcdService(cfg)
+	if _, err := client.Core().Services(metav1.NamespaceSystem).Create(&etcdService); err != nil {
+		return fmt.Errorf("failed to create self-hosted etcd service [%v]", err)
+	}
+	return nil
+}
+
+func launchSelfHostedProxy(cfg *kubeadmapi.MasterConfiguration, client *clientset.Clientset) error {
+	sa := getKubeProxyServiceAccount()
+	if _, err := client.CoreV1().ServiceAccounts(metav1.NamespaceSystem).Create(&sa); err != nil {
+		return fmt.Errorf("failed to create self-hosted kube-proxy serviceaccount [%v]", err)
```

Comment:
You mean add in retry logic? I just followed the exist process. If we want to add that I suggest a follow-up PR that does it for all self-hosted components.